// @thi.ng/math/int.js
const M8 = 0xff;
const M16 = 0xffff;
export const signExtend8 = (a) => ((a &= M8), a & 0x80 ? a | ~M8 : a);
export const signExtend16 = (a) => ((a &= M16), a & 0x8000 ? a | ~M16 : a);
export const addi8 = (a, b) => signExtend8((a | 0) + (b | 0));
export const divi8 = (a, b) => signExtend8((a | 0) / (b | 0));
export const muli8 = (a, b) => signExtend8((a | 0) * (b | 0));
export const subi8 = (a, b) => signExtend8((a | 0) - (b | 0));
export const andi8 = (a, b) => signExtend8((a | 0) & (b | 0));
export const ori8 = (a, b) => signExtend8(a | 0 | (b | 0));
export const xori8 = (a, b) => signExtend8((a | 0) ^ (b | 0));
export const noti8 = (a) => signExtend8(~a);
export const lshifti8 = (a, b) => signExtend8((a | 0) << (b | 0));
export const rshifti8 = (a, b) => signExtend8((a | 0) >> (b | 0));
export const addi16 = (a, b) => signExtend16((a | 0) + (b | 0));
export const divi16 = (a, b) => signExtend16((a | 0) / (b | 0));
export const muli16 = (a, b) => signExtend16((a | 0) * (b | 0));
export const subi16 = (a, b) => signExtend16((a | 0) - (b | 0));
export const andi16 = (a, b) => signExtend16((a | 0) & (b | 0));
export const ori16 = (a, b) => signExtend16(a | 0 | (b | 0));
export const xori16 = (a, b) => signExtend16((a | 0) ^ (b | 0));
export const noti16 = (a) => signExtend16(~a);
export const lshifti16 = (a, b) => signExtend16((a | 0) << (b | 0));
export const rshifti16 = (a, b) => signExtend16((a | 0) >> (b | 0));
export const addi32 = (a, b) => ((a | 0) + (b | 0)) | 0;
export const divi32 = (a, b) => ((a | 0) / (b | 0)) | 0;
export const muli32 = (a, b) => ((a | 0) * (b | 0)) | 0;
export const subi32 = (a, b) => ((a | 0) - (b | 0)) | 0;
export const andi32 = (a, b) => (a | 0) & (b | 0);
export const ori32 = (a, b) => a | 0 | (b | 0);
export const xori32 = (a, b) => (a | 0) ^ (b | 0);
export const lshifti32 = (a, b) => (a | 0) << (b | 0);
export const rshifti32 = (a, b) => (a | 0) >> (b | 0);
export const noti32 = (a) => ~a;
export const addu8 = (a, b) => ((a & M8) + (b & M8)) & M8;
export const divu8 = (a, b) => ((a & M8) / (b & M8)) & M8;
export const mulu8 = (a, b) => ((a & M8) * (b & M8)) & M8;
export const subu8 = (a, b) => ((a & M8) - (b & M8)) & M8;
export const andu8 = (a, b) => a & M8 & (b & M8) & M8;
export const oru8 = (a, b) => ((a & M8) | (b & M8)) & M8;
export const xoru8 = (a, b) => ((a & M8) ^ (b & M8)) & M8;
export const notu8 = (a) => ~a & M8;
export const lshiftu8 = (a, b) => ((a & M8) << (b & M8)) & M8;
export const rshiftu8 = (a, b) => ((a & M8) >>> (b & M8)) & M8;
export const addu16 = (a, b) => ((a & M16) + (b & M16)) & M16;
export const divu16 = (a, b) => ((a & M16) / (b & M16)) & M16;
export const mulu16 = (a, b) => ((a & M16) * (b & M16)) & M16;
export const subu16 = (a, b) => ((a & M16) - (b & M16)) & M16;
export const andu16 = (a, b) => a & M16 & (b & M16) & M16;
export const oru16 = (a, b) => ((a & M16) | (b & M16)) & M16;
export const xoru16 = (a, b) => ((a & M16) ^ (b & M16)) & M16;
export const notu16 = (a) => ~a & M16;
export const lshiftu16 = (a, b) => ((a & M16) << (b & M16)) & M16;
export const rshiftu16 = (a, b) => ((a & M16) >>> (b & M16)) & M16;
export const addu32 = (a, b) => ((a >>> 0) + (b >>> 0)) >>> 0;
export const divu32 = (a, b) => ((a >>> 0) / (b >>> 0)) >>> 0;
export const mulu32 = (a, b) => ((a >>> 0) * (b >>> 0)) >>> 0;
export const subu32 = (a, b) => ((a >>> 0) - (b >>> 0)) >>> 0;
export const andu32 = (a, b) => ((a >>> 0) & (b >>> 0)) >>> 0;
export const oru32 = (a, b) => ((a >>> 0) | (b >>> 0)) >>> 0;
export const xoru32 = (a, b) => ((a >>> 0) ^ (b >>> 0)) >>> 0;
export const notu32 = (a) => ~a >>> 0;
export const lshiftu32 = (a, b) => ((a >>> 0) << (b >>> 0)) >>> 0;
export const rshiftu32 = (a, b) => ((a >>> 0) >>> (b >>> 0)) >>> 0;
